codalab/model/bundle_model.py

```diff
@@ -986,11 +986,13 @@ def save_bundle(self, bundle):
             self.do_multirow_insert(connection, cl_bundle_metadata, metadata_values)
             bundle.id = result.lastrowid
 
-    def update_bundle(self, bundle, update, connection=None):
+    def update_bundle(self, bundle, update, connection=None, delete=False):
         """
-        Update a bundle's columns and metadata in the database and in memory.
-        The update is done as a diff: columns that do not appear in the update dict
+        Update a bundle's columns and metadata (default) or delete a bundle's metadata in the database and in memory.
+        1. The update is done as a diff: columns that do not appear in the update dict
         and metadata keys that do not appear in the metadata sub-dict are unaffected.
+        2. The delete is done as a diff on metadata dict only:
```

Comment:
I think using a `delete` flag to signal that only the keys in `update` to be used is confusing; usually, you want arguments to be as independent as possible, and here the semantics of the arguments are deeply intertwined.

One option is just to assume that keys with `None` values in `update` are deleted.  I *think* this should be safe, but you should double check if a null value is ever needed in the database apart from not existing.  Then you don't need `delete` at all as a flag.

Reply:
Yeah, I also thought about this several times during development and I do feel that adding `delete=True` is somehow confusing and redundant in the way it looks like right now. Original I was also wondering if I should directly set the specific field that I want to delete to be `None`, then using this as an indicator of deletion operation. However, I was thinking what if say a user originally set `request-queue=a`, later for some reason he wants to run his jobs on CodaLab's public workers, then he removes the corresponding value of `request-queue` from UI, meaning the `request-queue` now becomes `None`. With our deletion logic, the metadata key `request-queue` will be removed out from the metadata list and not shows up in UI eventually. Based on the definition of `request-queue` specified [here](https://github.com/codalab/codalab-worksheets/blob/master/codalab/bundles/run_bundle.py#L32), it is default set to `None`.

Reply:
I guess the function naming is a bit confusing as well, especially when describing a database operation. There are three types of operations when making changes to a table: `insert`, `update` and `delete`. We here merged `insert` with `update` as a single function. I think splitting these three operations in a certain way could definitely help improve the readability for this piece of code.